### DISCUSSION TITLE
Running SR on a distributed cluster

### DISCUSSION THREAD

**Discussion by chrhck** [MilesCranmer/PySR#644] [Q&A] [answered]
We have been SymbolicRegression.jl for our research but have hit a point where the equation search takes too long on a single compute node (with ~16 cores or so). We're now looking into using our distributed computing resources (we have both an MPI cluster as well as a slurm HPC cluster available) and were wondering if you have used SR in such an environment before or might know someone who has? we're hoping to not having to re-invent the wheel for writing the entire orchestration code (copy from slack DM)

**Comment by MilesCranmer** [comment, marked as answer]
Hey Christian,

Yes, I use SR like this in my own work. Basically if you just set `parallelism=:multiprocessing` then you can either:

1. Pass the process objects explicitly to the procs parameter (whether those procs are on the same node, or multiple nodes, etc.)
2. Or, set, for example, `numprocs=num_nodes * num_cores, addprocs_function=addprocs_slurm` , and SR.jl will try its best to set it up for you

I usually do the 2nd out of convenience

Just be sure to launch SR only once, on a single node, from a single task on the slurm job. ClusterManagers.jl will run srun internally for you.

**Reply by chrhck** [reply]
Thanks Miles. A followup question: We've been struggling to scale up SR to our cluster. We see very long idle times on the worker nodes at startup (up to a few hours) ![image](https://github.com/MilesCranmer/PySR/assets/43646111/61ce8926-c1c0-465d-8b33-f48cbba6975d)
 and after loading a previously saved state (up to 17h)
![image](https://github.com/MilesCranmer/PySR/assets/43646111/2a483086-d3e2-4379-bdee-0cc98f81d614)

We're running this on a slurm cluster with 64 nodes (20 CPU each).
Our dataset is very large (100k - 5000k datapoints). We've been experimenting with the batch-size, number of populations, and ncycles_per_iteration but so far have not found a combination that reduces the idle times. Any suggestions on what we could to to improve our resource utilization?

**Reply by MilesCranmer** [reply]
> Our dataset is very large (100k - 5000k datapoints)

This is much larger than is typically needed. Have you tried downsampling it? Symbolic regression is much less likely to overfit compared to other machine learning algorithms as it is not as expressive, so you can work with much fewer datapoints.

In fact the highest I will go in my own applications is only ~10k points. If my true dataset is 100k points I will just randomly subsample it to 10k. (And you can try to do importance sampling if your random subsampling does not capture all types of behavior observed in your dataset)

**Reply by chrhck** [reply]
Our problem is 5D and the data are pretty noisy - we can revisit subsampling the data again, but previously we found that we get better results with the full dataset. If we have to remain with the full dataset - is there anything we can do to improve performance?

**Reply by MilesCranmer** [reply]
If the data is noisy, one option that I've used in the past is to first fit a flexible ML model to the data first, with an MSE loss, which acts to average out the noise. Usually I will use a Gaussian Process with some flexible kernel, like the built-in one here:
https://github.com/MilesCranmer/PySR/blob/89e991d9d7a5e34025bdbbe9076a8fd0892f04cc/pysr/denoising.py#L9-L28

However, GPs scale poorly with number of points... Even 100k will be horrendously slow. So you might want to use a multi-layer perceptron or XGBoost[^1].

Fit those to the data first, then generate and evaluate on some grid of inputs. The result should have less noise (though there might be biases introduced here, so do this with care).

Then, fit PySR on those "noiseless" samples. With only 5 features you probably can get a decent result with <1k points.

And with that result you can evaluate on the larger dataset to evaluate it.

[^1]: Speaking of which, maybe PySR's `denoise` option should default to `XGBoost` when there is too much data...

**Reply by dhjx1996** [reply]
> Hey Christian,
> 
> Yes, I use SR like this in my own work. Basically if you just set `parallelism=:multiprocessing` then you can either:
> 
> 1. Pass the process objects explicitly to the procs parameter (whether those procs are on the same node, or multiple nodes, etc.)
> 2. Or, set, for example, `numprocs=num_nodes * num_cores, addprocs_function=addprocs_slurm` , and SR.jl will try its best to set it up for you
> 
> I usually do the 2nd out of convenience
> 
> Just be sure to launch SR only once, on a single node, from a single task on the slurm job. ClusterManagers.jl will run srun internally for you.

Hi @MilesCranmer, may I clarify where or how do I specify `addprocs_function=addprocs_slurm`? Alternatively, could you provide a simple example for running PySR over multiple nodes? My workflow is `model = PySRRegressor(**binary_operators and other options**)` followed by `model.fit(X, y)`.

**Reply by MilesCranmer** [reply]
In PySR the equivalent is to use `parallelism="multiprocessing", cluster_manager="slurm"`. (Set the number of procs too)

Just run this once on the head node and it will do the `srun` call internally.

**Reply by dhjx1996** [reply]
Oh, my impression is that [ClusterManagers.jl](https://github.com/JuliaParallel/ClusterManagers.jl) is not working (I read through their issues page and saw your comments). In any case when I tried `cluster_manager="slurm"` I got the error:

_juliacall.JuliaError: ArgumentError: Package ClusterManagers not found in current path. - Run `import Pkg; Pkg.add("ClusterManagers")` to install the ClusterManagers package._

**Full traceback:**

Traceback (most recent call last):
  File "/burg/home/dh3065/twostream_PySR_RTsources_Dec06_small5_multinode/twostream_PySR_RTsources.py", line 127, in <module>
    modelRpT.fit(X_sample, y1_sample)
  File "/burg/home/dh3065/miniconda3/envs/2stream/lib/python3.12/site-packages/pysr/sr.py", line 2087, in fit
    self._run(X, y, runtime_params, weights=weights, seed=seed)
  File "/burg/home/dh3065/miniconda3/envs/2stream/lib/python3.12/site-packages/pysr/sr.py", line 1694, in _run
    cluster_manager = _load_cluster_manager(cluster_manager)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/burg/home/dh3065/miniconda3/envs/2stream/lib/python3.12/site-packages/pysr/julia_helpers.py", line 31, in _load_cluster_manager
    jl.seval(f"using ClusterManagers: addprocs_{cluster_manager}")
  File "/burg/home/dh3065/.julia/packages/PythonCall/Nr75f/src/JlWrap/module.jl", line 27, in seval
    return self._jl_callmethod($(pyjl_methodnum(pyjlmodule_seval)), expr)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
juliacall.JuliaError: ArgumentError: Package ClusterManagers not found in current path.
- Run `import Pkg; Pkg.add("ClusterManagers")` to install the ClusterManagers package.
Stacktrace:
  [1] macro expansion
    @ ./loading.jl:1772 [inlined]
  [2] macro expansion
    @ ./lock.jl:267 [inlined]
  [3] __require(into::Module, mod::Symbol)
    @ Base ./loading.jl:1753
  [4] #invoke_in_world#3
    @ ./essentials.jl:926 [inlined]
  [5] invoke_in_world
    @ ./essentials.jl:923 [inlined]
  [6] require(into::Module, mod::Symbol)
    @ Base ./loading.jl:1746
  [7] eval
    @ ./boot.jl:385 [inlined]
  [8] eval
    @ ./Base.jl:88 [inlined]
  [9] pyjlmodule_seval(self::Module, expr::Py)
    @ PythonCall.JlWrap ~/.julia/packages/PythonCall/Nr75f/src/JlWrap/module.jl:13
 [10] _pyjl_callmethod(f::Any, self_::Ptr{PythonCall.C.PyObject}, args_::Ptr{PythonCall.C.PyObject}, nargs::Int64)
    @ PythonCall.JlWrap ~/.julia/packages/PythonCall/Nr75f/src/JlWrap/base.jl:67
 [11] _pyjl_callmethod(o::Ptr{PythonCall.C.PyObject}, args::Ptr{PythonCall.C.PyObject})
    @ PythonCall.JlWrap.Cjl ~/.julia/packages/PythonCall/Nr75f/src/JlWrap/C.jl:63

Any help would be much appreciated! Also let me know if I should move this discussion to a new thread.

**Reply by MilesCranmer** [reply]
Yes please post an issue: https://github.com/MilesCranmer/PySR/issues/new/choose

**Reply by MilesCranmer** [reply]
(I saw that ClusterManagers.jl might be working again – I think it was just out-of-date for a given slurm version or something)

**Reply by dhjx1996** [reply]
> Yes please post an issue: https://github.com/MilesCranmer/PySR/issues/new/choose

I have created an issue, thank you very much for your help.
